torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -592,7 +593,7 @@ void Reducer::mark_bucket_ready(size_t bucket_index) {
     // TODO(@sinannasir): merge `work` and `future_work`. Related to GH Issue
     // #41266.
     if (comm_hook_ == nullptr) {
-      bucket.work = process_group_->allreduce(tensors);
+      bucket.future_work = process_group_->allreduce(tensors)->getFuture();
```

Comment:
We can get rid of this if-else completely and register a hook in DDP constructor that does allreduce by default? Then, we should allow users to override the hook (I think we only allow the hook to be registered once) as many times as they wish.

Reply:
Included a new class named `AllreduceHook` and registered that as default comm hook.